src/interface/data_collection.cpp

```diff
@@ -54,25 +54,51 @@ std::shared_ptr<T> &DataCollection<T>::AddShallow(const std::string &label,
   return Add(label, src, flags, true);
 }
 
-template <>
 std::shared_ptr<MeshData<Real>> &
-DataCollection<MeshData<Real>>::GetOrAdd(const std::string &mbd_label,
-                                         const int &partition_id) {
-  const std::string label = mbd_label + "_part-" + std::to_string(partition_id);
+GetOrAdd_impl(Mesh *pmy_mesh_,
+              std::map<std::string, std::shared_ptr<MeshData<Real>>> &containers_,
+              BlockList_t &block_list, const std::string &mbd_label,
+              const int &partition_id, const int gmg_level) {
+  std::string label = mbd_label + "_part-" + std::to_string(partition_id);
+  if (gmg_level >= 0) label = label + "_gmg-" + std::to_string(gmg_level);
```

Comment:
IIRC the logical level can become negative, correct? And I also see below that for standard containers this is set to -1, also correct?
So how would the code know how to differentiate between a standard container and a GMG container with a negative level?

Reply:
`gmg_level` should be in the range [0, total number of gmg levels) so that zero corresponds to the most negative logical level.

Reply:
Gotcha, so logical level and gmg level are to different things.